### DISCUSSION TITLE
Where is `pager_print` implemented?

### DISCUSSION THREAD

**Discussion by evbernardes** [sympy/sympy#25084] [Q&A] [answered]
I'm implementing a class for dual quaternions that is heavily based on the implementation from the `Quaternion` class.

I implemented both a simple printer and a latex printer, works as expected.

Some function calls lead to this error though:
```python
AttributeError: module 'sympy' has no attribute 'DualQuaternionpager_print'
```

I decided to do a search for `Quaternionpager_print`, to see where this should be re-implemented for my class, but I wasn't able to find anything. Does anyone know what I'm doing wrong?

Full traceback:
```python
Traceback (most recent call last):

  Cell In[65], line 1
    1 * Q1

  File ~/.softwares/sympy/sympy/algebras/dual_quaternion.py:453 in __rmul__
    return self._generic_mul(_sympify(other), self)

  File ~/.softwares/sympy/sympy/algebras/dual_quaternion.py:626 in _generic_mul
    return DualQuaternion(re(Q1), im(Q1), 0, 0) * Q2

  File ~/.softwares/sympy/sympy/algebras/dual_quaternion.py:450 in __mul__
    return self._generic_mul(self, _sympify(other))

  File ~/.softwares/sympy/sympy/algebras/dual_quaternion.py:643 in _generic_mul
    return DualQuaternion.from_quaternions(q, p)

  File ~/.softwares/sympy/sympy/algebras/dual_quaternion.py:213 in from_quaternions
    return cls(*non_dual.args, *dual.args, norm=non_dual.norm)

  File ~/.softwares/sympy/sympy/algebras/dual_quaternion.py:85 in __new__
    obj._q.set_norm(norm)

  File ~/.softwares/sympy/sympy/algebras/quaternion.py:157 in set_norm
    norm = sympify(norm)

  File ~/.softwares/sympy/sympy/core/sympify.py:495 in sympify
    expr = parse_expr(a, local_dict=locals, transformations=transformations, evaluate=evaluate)

  File ~/.softwares/sympy/sympy/parsing/sympy_parser.py:1049 in parse_expr
    exec('from sympy import *', global_dict)

  File <string>:1

AttributeError: module 'sympy' has no attribute 'DualQuaternionpager_print'
```

**Comment by sylee957** [comment]
There is `pager_print` in
https://github.com/sympy/sympy/blob/bad2d0541fd260f3843b695c0d57ee3cde0756bb/sympy/printing/pretty/pretty.py#L2931-L2946
But I can't see where `'DualQuaternionpager_print'` came from unless you can share the source code.

**Reply by evbernardes** [reply]
@sylee957 I think there is something about `from sympy import *` that breaks it, and there is something internal to `sympy` making this star import. 

I made another simpler example to demonstrate it. I created a `sympy/algebras/unit_quaternion.py` file that defines a very simple class:

```python
from sympy.algebras.quaternion import Quaternion

class UnitQuaternion(Quaternion):
    _op_priority = 11.0

    is_commutative = False

    def __new__(cls, a=0, b=0, c=0, d=0, real_field=True):
        obj = Quaternion.__new__(cls, a, b, c, d, real_field)
        return obj

    def norm(self):
        return 1
```

In `sympy/algebras/__init__.py` I added:
```python
from .quaternion import Quaternion
from .unit_quaternion import UnitQuaternion

__all__ = ["Quaternion", "UnitQuaternion"]
```

And in `sympy/__init__.py`:
```python
from .algebras import Quaternion, UnitQuaternion
```
... and:
```python
    # sympy.algebras
    'Quaternion', 'UnitQuaternion'
```

I made a call to this new class and it works as expected, calling the printers of `Quaternion`. Now, trying to do `from sympy import *` gives me the same kind of error:

```python
  Cell In[11], line 1
    from sympy import *

AttributeError: module 'sympy' has no attribute 'UnitQuaternionpager_print'
```

**Reply by sylee957** [reply, marked as answer]
I think that that is likely a mistake in your code that you don't put comma and the string gets concatenated

```
    # sympy.algebras
    'Quaternion', 'UnitQuaternion',
```

`pager_print` is at next line, so this is likely how it got wrong

**Reply by evbernardes** [reply]
Oh my, thank you for this! Definitely on my top 5 dumbest bugs! :D
